File: python/common/r2.py
"""Cloudflare R2 (S3-compatible) 클라이언트 래퍼."""
from __future__ import annotations

import io
import os
from dataclasses import dataclass
from typing import Iterator

import boto3
import pandas as pd
import pyarrow.parquet as pq


@dataclass
class R2Config:
    endpoint: str
    bucket: str
    access_key: str
    secret_key: str
    region: str = "auto"

    @classmethod
    def from_env(cls) -> "R2Config":
        return cls(
            endpoint=os.environ["R2_ENDPOINT"],
            bucket=os.environ.get("R2_BUCKET", "lemuel-quant"),
            access_key=os.environ["R2_ACCESS_KEY"],
            secret_key=os.environ["R2_SECRET_KEY"],
        )


class R2Client:
    """R2 위에 올라간 data-warehouse Parquet 스냅샷을 읽기 위한 헬퍼."""

    def __init__(self, cfg: R2Config | None = None):
        self.cfg = cfg or R2Config.from_env()
        self._s3 = boto3.client(
            "s3",
            endpoint_url=self.cfg.endpoint,
            region_name=self.cfg.region,
            aws_access_key_id=self.cfg.access_key,
            aws_secret_access_key=self.cfg.secret_key,
        )

    def list_keys(self, prefix: str = "snapshots/") -> list[str]:
        keys: list[str] = []
        token: str | None = None
        while True:
            kwargs = {"Bucket": self.cfg.bucket, "Prefix": prefix}
            if token:
                kwargs["ContinuationToken"] = token
            resp = self._s3.list_objects_v2(**kwargs)
            for obj in resp.get("Contents", []):
                keys.append(obj["Key"])
            if not resp.get("IsTruncated"):
                break
            token = resp.get("NextContinuationToken")
        return keys

    def read_parquet(self, key: str) -> pd.DataFrame:
        body = self._s3.get_object(Bucket=self.cfg.bucket, Key=key)["Body"].read()
        return pq.read_table(io.BytesIO(body)).to_pandas()

    def read_many(self, keys: list[str]) -> Iterator[pd.DataFrame]:
        for k in keys:
            yield self.read_parquet(k)

File: r/common/r2.R
# Cloudflare R2 (S3 호환) 클라이언트. data-warehouse 산출 Parquet 읽기 전용.
#
# 사용:
#   source("r/common/r2.R")
#   df <- r2_read_parquet("snapshots/rollup-20260507-100000.parquet")
#   keys <- r2_list("snapshots/")

suppressPackageStartupMessages({
  library(paws.storage)
  library(arrow)
  library(tibble)
})

r2_client <- function(cfg = NULL) {
  if (is.null(cfg)) {
    cfg <- list(
      endpoint   = Sys.getenv("R2_ENDPOINT"),
      bucket     = Sys.getenv("R2_BUCKET", "lemuel-backup"),
      access_key = Sys.getenv("R2_ACCESS_KEY"),
      secret_key = Sys.getenv("R2_SECRET_KEY"),
      region     = "auto"
    )
  }
  stopifnot(nzchar(cfg$endpoint), nzchar(cfg$access_key), nzchar(cfg$secret_key))
  s3 <- paws.storage::s3(
    config = list(
      endpoint    = cfg$endpoint,
      region      = cfg$region,
      credentials = list(creds = list(
        access_key_id     = cfg$access_key,
        secret_access_key = cfg$secret_key
      ))
    )
  )
  structure(list(s3 = s3, bucket = cfg$bucket), class = "r2_client")
}

r2_list <- function(prefix = "snapshots/", client = r2_client()) {
  keys <- character()
  token <- NULL
  repeat {
    args <- list(Bucket = client$bucket, Prefix = prefix)
    if (!is.null(token)) args$ContinuationToken <- token
    resp <- do.call(client$s3$list_objects_v2, args)
    keys <- c(keys, vapply(resp$Contents, `[[`, character(1), "Key"))
    if (!isTRUE(resp$IsTruncated)) break
    token <- resp$NextContinuationToken
  }
  keys
}

r2_read_parquet <- function(key, client = r2_client()) {
  resp <- client$s3$get_object(Bucket = client$bucket, Key = key)
  arrow::read_parquet(arrow::BufferReader$new(resp$Body))
}

# trade 채널 데이터를 종목별로 모아 단일 tibble 로 반환.
r2_load_trades <- function(symbol,
                            channel_prefix = "trade.binance",
                            client = r2_client()) {
  keys <- r2_list("snapshots/", client = client)
  frames <- list()
  for (k in keys) {
    df <- tryCatch(r2_read_parquet(k, client = client), error = function(e) NULL)
    if (is.null(df) || !nrow(df)) next
    if (!"channel" %in% names(df)) next
    sub <- df[df$channel == "trade", , drop = FALSE]
    if (!nrow(sub) || !"tag.symbol" %in% names(sub)) next
    sub <- sub[sub$`tag.symbol` == symbol, , drop = FALSE]
    if ("tag.channel" %in% names(sub)) {
      sub <- sub[startsWith(sub$`tag.channel`, channel_prefix), , drop = FALSE]
    }
    if (!nrow(sub)) next
    frames[[length(frames) + 1L]] <- tibble::as_tibble(sub)
  }
  if (!length(frames)) return(tibble::tibble())
  out <- do.call(rbind, frames)
  out$ts <- as.POSIXct(out$ts_ns / 1e9, origin = "1970-01-01", tz = "UTC")
  out
}
